cql3/restrictions/restriction.hh

```diff
@@ -49,15 +50,87 @@
 #include "cql3/term.hh"
 #include "cql3/statements/bound.hh"
 #include "index/secondary_index_manager.hh"
+#include "query-result-reader.hh"
 #include "types.hh"
 
 namespace cql3 {
 
+namespace selection {
+class selection;
+} // namespace selection
+
 namespace restrictions {
 
+class restriction;
+
 struct allow_local_index_tag {};
 using allow_local_index = bool_class<allow_local_index_tag>;
 
+/// Work-in-progress new restriction representation.  When complete, it will replace the old one.
+namespace wip {
+
+// The new representation exposes its member data publicly.  Operations on restrictions are
+// performed by free functions that take restrictions as parameters and use the visitor pattern to
+// specialize code for different kinds of restrictions.
+//
+// The most interesting class is wip::binary_operator below, which can represent both multi- and
+// single-column restrictions.  Instead of the old merge_with mechanism for combining restrictions,
+// we will simply add them into a conjunction expression that will be processed using visitors.
+
+class binary_operator;
+class conjunction;
+
+/// A restriction expression -- union of all possible restriction types.  bool means a Boolean constant.
+using expression = std::variant<bool, conjunction, binary_operator>;
+
+/// A column, optionally subscripted by a term (eg, c1 or c2['abc']).
+struct column_value {
+    const column_definition* col;
+    ::shared_ptr<term> sub; ///< If present, this LHS is col[sub], otherwise just col.
+    /// For easy creation of vector<column_value> from vector<column_definition*>.
+    column_value(const column_definition* col) : col(col) {}
+    /// The compiler doesn't auto-generate this due to the other constructor's existence.
+    column_value(const column_definition* col, ::shared_ptr<term> sub) : col(col), sub(sub) {}
+};
+
+/// Represents token function on LHS of an operator relation.  No need to list column definitions
+/// here -- token takes exactly the partition key as its argument.
+struct token {};
+
+/// Operator restriction: LHS op RHS.
+struct binary_operator {
+    std::variant<std::vector<column_value>, token> lhs;
+    const operator_type* op; // Pointer because operator_type isn't copyable or assignable.
+    ::shared_ptr<term> rhs;
```

Comment:
Why are right hand side and left hand side allowed types  reduced to column values and terms? Will it ever be possible to do expr < expr?

Reply:
> Why are right hand side and left hand side allowed types reduced to column values and terms?

That's just to limit the scope of changes in this step.  The upstream code that creates the expression will reject any other subexpression types anyway.

> Will it ever be possible to do expr < expr?

Yes, after the old code is fully replaced.  At that time, LHS and RHS will simply become another expression.